src/osd/PrimaryLogPG.cc

```diff
@@ -7110,6 +7185,45 @@ void PrimaryLogPG::start_copy(CopyCallback *cb, ObjectContextRef obc,
   _copy_some(obc, cop);
 }
 
+void PrimaryLogPG::start_repair_copy(OpContext *ctx, uint32_t what,
+				     const vector<pg_shard_t>& bad_shards)
+{
+  const auto& dest = ctx->obs->oi.soid;
+  auto found = copy_ops.find(dest);
+  if (found != copy_ops.end()) {
+    cancel_copy(found->second, false);
+  }
+  // we expect the client side to prepend an assert_version() call before
+  // this op, so just pass 0 as the version.
+  auto cb = new CopyFromCallback(ctx);
+  ctx->copy_cb = cb;
+  uint32_t to_copy = 0;
+  if (what & librados::repair_copy_t::DATA)
```

Comment:
i guess it's fine at this moment: the internal representation can share with the same definitions defined in public interface. and i think it's desirable. if they diverge over the time, we certainly have the luxury to use another set of protocol internally and do the translation in librados. and we are also using other stuff in the `librados` in `PrimaryLogPG.cc` if you search `librados` in that source file.